src/login/logind-session.c

```diff
@@ -1126,15 +1143,21 @@ static void session_restore_vt(Session *s) {
 
         int vt, old_fd;
 
-        /* We need to get a fresh handle to the virtual terminal,
-         * since the old file-descriptor is potentially in a hung-up
-         * state after the controlling process exited; we do a
-         * little dance to avoid having the terminal be available
-         * for reuse before we've cleaned it up.
-         */
+        /* We do a little dance to avoid having the terminal be available for
+         * reuse before we've cleaned it up. */
         old_fd = TAKE_FD(s->vtfd);
 
-        vt = session_open_vt(s);
+        /* The virtual terminal can potentially be entering in hung-up state at
+         * any time depending on when the controlling process exits.
+         *
+         * If the controlling process exits while we're restoring the virtual
+         * terminal, the VT will enter in hung-up state and we'll fail at doing
+         * its re-initialization. To prevent this case, we forcibly become the
+         * controlling process of the VT and release it when we're done.
+         *
+         * If the controlling process already exited, getting a fresh handle to
+         * the virtual terminal reset the hung-up state. */
+        vt = session_acquire_vt(s);
```

Comment:
hmm, not sure how I feel about this. being the controlling process of a tty means a lot of things, including signal delivery due to C-c and things like this. I am really sure logind should not become controlling process of any tty, neither for short nor for long.

What's the precise reason for becoming controlling process? I don't get this? isn't it sufficient to kick off however else is on it? is there no proper API for doing that? if not, I think it would make more sense to fork off a short-lived child process that acquires the tty and then exits, with logind waiting for it. but logind itself should not play around with tty ownership for itself I am sure